lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp

```diff
@@ -1374,38 +1450,45 @@ LinearLayout chooseScaledWmmaScaleLayout(MLIRContext *ctx, int dotOperandIdx,
 
-  // If there's 1 tile per warp, we are not using the remaining 16 lanes, so
-  // just let them duplicate values of the first 16 lanes.
-  // Otherwise, we put consecutive values along the nonK dim in the remaining
-  // 16 lanes.
   unsigned mnDim = dotOperandIdx == 0 ? rank - 2 : rank - 1;
-  unsigned tilePerWarpMN = tilesPerWarp[mnDim];
-  if (tilePerWarpMN > 1) {
-    assert(tilePerWarpMN == 2 && "TilesPerWarp > 2 is not supported.");
-    tileLayout *= LinearLayout::identity1D(tilePerWarpMN, kLane, dimNonK);
-  } else {
-    tileLayout *= LinearLayout::zeros1D(2, kLane, dimNonK);
-  }
 
   // If the shape along the K dim is larger than kWidth, repeat this
   // pattern to fill the K dim.
   tileLayout *= LinearLayout::identity1D(kSize / scaleKWidth, kRegister, dimK);
 
-  auto warpsPerCTANew = (dotOperandIdx == 1)
-                            ? SmallVector{warpsPerCTA[1], warpsPerCTA[0]}
-                            : SmallVector{warpsPerCTA[0], warpsPerCTA[1]};
+  auto firstRepInNonK = tileLayout.getInDimSizeLog2(kRegister);
 
-  auto warpOrder = (dotOperandIdx == 1) ? SmallVector<unsigned>{0, 1}
-                                        : SmallVector<unsigned>{1, 0};
-  LinearLayout warpLayout =
-      identityStandardND(kWarp, warpsPerCTANew, warpOrder);
-  LinearLayout ctaLayout = tileLayout.transposeOuts(outDimNames) *
-                           warpLayout.transposeOuts(outDimNames);
+  if (dotOperandIdx == 1) {
+    warpMfmaLayout = transposeLinearLayout(warpMfmaLayout, order);
+  }
 
-  return combineCtaCgaWithShape(
-      ctaLayout, CGAEncodingAttr::getDefault(ctx, /*rank=*/2), dotOperandShape);
+  // Zero out M or N dim based on opIdx
+  auto warpDotMfmaLayout = projectAwayOutDim(warpMfmaLayout, dimK, ctx);
+  // If repetition (aka register basis) iz 0 in all out dims we need to remove
+  // it since this repetition doesn't make sense for dotOp layout.
+  warpDotMfmaLayout =
+      actionRemoveBroadcastedRegs(warpDotMfmaLayout).apply(warpDotMfmaLayout);
+
+  LinearLayout ctaLayout = tileLayout;
+  ctaLayout = tileLayout.transposeOuts(outDimNames) * warpDotMfmaLayout;
+  auto nonOpSelLayout = combineCtaCgaWithShape(
+      ctaLayout, CGAEncodingAttr::getDefault(ctx, /*rank=*/rank),
+      dotOperandShape);
+
+  // This is the tricky part. For a single tile, only 16 threads
+  // hold scale values, 4 for each thread. Other 16 thread in a warp
+  // broadcast these values. This is a waste of memory. In order to deal with
+  // that we can assignd other 16 threads (thread 15-31), to hold scales of the
+  // next tile computed by the same warp (aka it's first repetition in non-k
+  // dim), if there is one. So register base that naturally representes first
+  // repetition needs to be moved to lane base that representes lane 16. Since
+  // for a single tile thread holds 4 vals, we move register base 2, to lane
+  // base 4.
+  auto retLayout = replaceFirstRepWithLane(nonOpSelLayout, firstRepInNonK, ctx);
+  return retLayout;
```

Comment:
Why do we need this now and we didn't need it before? Also, shouldn't this be part of the tile and not a patch-up?